### ISSUE TITLE
 Performance of Include() on RC2

### ISSUE THREAD

**Issue by janiceattard** [dotnet/efcore#5530]
Having the below model:

```
public partial class Products
 {
        public Products()
        {
        }

        public int Id { get; set; }
        ...
        public int? DocId { get; set; }
        public virtual Documents Doc { get; set; }
}

public partial class Documents
{
        public Documents()
        {
            Products = new HashSet<Products>();
        }

        public int Id { get; set; }
        ...
        public virtual ICollection<Products> Products { get; set; }
}
```

The SQL generated by the following queries is similar and both have similar timings in SSMS (< 0ms) and execution plans:

```
var query = from p in dbContext.Products
     where p.ProductId == req.ProductId
               && p.SiteId == req.Engine.Context.SiteId                             
     select p;

query = query.Include(b => b.Doc);
var x = query.ToList();


var query = from p in dbContext.Products
     join d in dbContext.Documents on p.DocId equals  d.Id
     where p.ProductId == req.ProductId
                && p.SiteId == req.Engine.Context.SiteId
     select new { p, d };

var x = query.ToList();
```

However, the when running the code the query with the Include() is very slow compared to the one with the join. 

With the Include() we are only able to run 54,500 queries in 1 minute
With the join we get 212,500 queries in 1 minute

**Comment by smitpatel** [comment]
@anpete - How did `Tweaks to warnings configuration` fixed this? Or is it just wrong issue number used?

**Comment by anpete** [comment]
@smitpatel Typo

**Comment by anpete** [comment]
@janiceattard Can you post the SQL generated for each query?

**Comment by janiceattard** [comment]
Query generated when using Include()

```
exec sp_executesql N'SELECT [p].[Id], [p].[DocId], [p].[ProductId], [p].[SiteId], [d].[Id], [d].[Country], [d].[DocumentId]
FROM [rdt].[Products] AS [p]
LEFT JOIN [rdt].[Documents] AS [d] ON [p].[DocId] = [d].[Id]
WHERE ([p].[ProductId] = @__req_ProductId_0) AND ([p].[SiteId] = @__req_Engine_Context_SiteId_1)',N'@__req_ProductId_0 varchar(8000),@__req_Engine_Context_SiteId_1 int',@__req_ProductId_0='324509925',@__req_Engine_Context_SiteId_1=1
```

Query generated when using join

```
exec sp_executesql N'SELECT [p].[Id], [p].[DocId], [p].[ProductId], [p].[SiteId], [d].[Id], [d].[Country], [d].[DocumentId]
FROM [rdt].[Products] AS [p]
INNER JOIN [rdt].[Documents] AS [d] ON [p].[DocId] = [d].[Id]
WHERE ([p].[ProductId] = @__req_ProductId_0) AND ([p].[SiteId] = @__req_Engine_Context_SiteId_1)',N'@__req_ProductId_0 varchar(8000),@__req_Engine_Context_SiteId_1 int',@__req_ProductId_0='324509925',@__req_Engine_Context_SiteId_1=1
```

**Comment by maumar** [comment]
@janiceattard It seems that Product-Documents is a optional navigation. This means that performing INNER JOIN is not safe as it will not return products that don't have a document associated with them

**Comment by janiceattard** [comment]
@maumar Yes, indeed. That is why we want to use the Include() but we noticed that when we upgraded from RC1 to RC2 the amount of queries per minute dropped drastically in cases where we were using Include().

**Comment by rowanmiller** [comment]
Can you confirm if the generated SQL was the same in RC1 and RC2

**Comment by janiceattard** [comment]
Yes, same SQL was generated for both RC1 and RC2.

**Comment by rowanmiller** [comment]
Ran the following code listing and the difference is because the related data is not loaded in RC1. In RC1, the returned `Product` has `Doc` set to null. In RTM the `Doc` property is correctly populated with data from the database. This explains why the RC1 query executes quicker 😄.

```c#
using Microsoft.EntityFrameworkCore;
using System;
using System.Collections.Generic;
using System.Linq;

namespace Repro.RC1.Full
{
    class Program
    {
        static void Main(string[] args)
        {
            using (var context = new TestContext())
            {
                Console.WriteLine("Creating database...");
                context.Database.EnsureDeleted();
                context.Database.EnsureCreated();

                Console.WriteLine("Inserting data...");
                context.Products.Add(new Products
                {
                    SiteId = 1,
                    Doc = new Documents()
                });
                context.SaveChanges();
            }

            using (var context = new TestContext())
            {
                var productId = 1;
                var siteId = 1;
                var count = 0;

                Console.WriteLine("Starting test...");
                var end = DateTime.Now.AddSeconds(30);

                while (DateTime.Now < end)
                {
                    var query = from p in context.Products
                                where p.Id == productId && p.SiteId == siteId
                                select p;

                    query = query.Include(b => b.Doc);

                    var x = query.ToList();

                    count++;
                }

                Console.WriteLine($"Result: {count} queries");
            }
        }
    }

    public class TestContext : DbContext
    {
        public DbSet<Products> Products { get; set; }
        public DbSet<Documents> Documents { get; set; }

        protected override void OnConfiguring(DbContextOptionsBuilder optionsBuilder)
        {
            optionsBuilder.UseSqlServer(@"Server=(localdb)\mssqllocaldb;Database=TestDb;Trusted_Connection=True;");
        }
    }


    public partial class Products
    {
        public int Id { get; set; }
        public int SiteId { get; set; }

        public int? DocId { get; set; }
        public virtual Documents Doc { get; set; }
    }

    public partial class Documents
    {
        public int Id { get; set; }

        public virtual ICollection<Products> Products { get; set; } = new HashSet<Products>();
    }
}
```

**Comment by janiceattard** [comment]
Hi,

Just tested again with RC1 to confirm that `Doc` was indeed null - however this was not the case as can be seen from the below screenshot:

![2016-07-08 10_33_40-doc png - windows photo viewer](https://cloud.githubusercontent.com/assets/17545705/16681711/ac40d538-44f6-11e6-9c6e-fdea993f3652.png)

_project.json_ used for RC1

```
{
  "version": "1.0.0-*",
  "tags": [ "" ],
  "projectUrl": "",
  "licenseUrl": "",
  "frameworks": {    
    "dnx451": {
      "dependencies": {
        
        "Microsoft.CSharp": "4.0.1-beta-23516",
        "System.Collections": "4.0.11-beta-23516",
        "System.Linq": "4.0.1-beta-23516",
        "System.Runtime": "4.0.21-beta-23516",
        "System.Threading": "4.0.11-beta-23516"
      }
    }
  },
  "dependencies": {
    "Model": "1.0.0-*",
    "Offers": "1.0.0-*",
    "Products": "1.0.0-*",
    "GeneralData": "1.0.0-*",
    "EntityFramework": "6.1.3",
    "Microsoft.Extensions.PlatformAbstractions": "1.0.0-rc1-final",
    "Microsoft.Extensions.Configuration": "1.0.0-rc1-final",
    "Microsoft.Extensions.Configuration.EnvironmentVariables": "1.0.0-rc1-final",
    "Microsoft.Extensions.Configuration.Json": "1.0.0-rc1-final",
    "NEST": "2.2.0"
  },
  "commands": {
    "ef": "EntityFramework.Commands"
  }
}
```

**Comment by rowanmiller** [comment]
@janiceattard can you take my code listing and modify it to something that demonstrates the performance difference you are seeing? I tried to reverse engineer your scenario based on the info in this issue, but there must be something different between our code.

**Comment by janiceattard** [comment]
Below are the two code listings - one for RC1 and one for RC2 with screenshots which show the preformance difference.

### RC1

_Program.cs_

```
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using Microsoft.Data.Entity;

namespace PerfIssue.RC1
{
    class Program
    {
        static void Main(string[] args)
        {
            Console.WriteLine("RC1");
            using (var context = new TestContext())
            {
                Console.WriteLine("Creating database...");
                context.Database.EnsureDeleted();
                context.Database.EnsureCreated();

                Console.WriteLine("Inserting data...");

                context.Documents.Add(new Documents { });

                context.Products.Add(new Products
                {
                    SiteId = 1,
                    Doc = new Documents
                    {
                        Id = 1
                    }
                });
                context.SaveChanges();
            }

            using (var context = new TestContext())
            {
                var productId = 1;
                var siteId = 1;
                var count = 0;

                Console.WriteLine("Starting test...");
                var end = DateTime.Now.AddSeconds(30);

                while (DateTime.Now < end)
                {
                    var query = from p in context.Products
                                where p.Id == productId && p.SiteId == siteId
                                select p;

                    query = query.Include(b => b.Doc);

                    var x = query.ToList();

                    count++;
                }

                Console.WriteLine($"Result: {count} queries");
                Console.ReadLine();
            }
        }
    }

    public class TestContext : DbContext
    {
        public DbSet<Products> Products { get; set; }
        public DbSet<Documents> Documents { get; set; }

        protected override void OnConfiguring(DbContextOptionsBuilder optionsBuilder)
        {
            optionsBuilder.UseSqlServer(@"Server=(localdb)\mssqllocaldb;Database=TestDb;Trusted_Connection=True;");
        }
    }


    public partial class Products
    {
        public int Id { get; set; }
        public int SiteId { get; set; }

        public int? DocId { get; set; }
        public virtual Documents Doc { get; set; }
    }

    public partial class Documents
    {
        public int Id { get; set; }

        public virtual ICollection<Products> Products { get; set; } = new HashSet<Products>();
    }
}

```

_project.json_
```
{
  "version": "1.0.0-*",
  "buildOptions": {
    "emitEntryPoint": true
  },

  "dependencies": {
    "EntityFramework.Commands": "7.0.0-rc1-final",
    "EntityFramework.MicrosoftSqlServer": "7.0.0-rc1-final",
    "EntityFramework.MicrosoftSqlServer.Design": "7.0.0-rc1-final",
    "EntityFramework.Core": "7.0.0-rc1-final",
    "Microsoft.Extensions.Logging": "1.0.0-rc1-final",
    "Microsoft.Extensions.Logging.Abstractions": "1.0.0-rc1-final",
    "Microsoft.Extensions.DependencyInjection.Abstractions": "1.0.0-rc1-final",
    "Microsoft.Extensions.DependencyInjection": "1.0.0-rc1-final"
  },

  "frameworks": {
    "dnx451": {
      "dependencies": {
        "Microsoft.CSharp": "4.0.1-beta-23516",
        "System.Collections": "4.0.11-beta-23516",
        "System.Linq": "4.0.1-beta-23516",
        "System.Runtime": "4.0.21-beta-23516",
        "System.Threading": "4.0.11-beta-23516"
      }
    }
  }
}
```

_Results_

55989 queries
![2016-07-11 12_04_36-lrpublisherweb - remote desktop connection](https://cloud.githubusercontent.com/assets/17545705/16727390/284077e4-4761-11e6-97d4-b87540d2c7c8.png)


### RC2

_Program.cs_

```
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using Microsoft.EntityFrameworkCore;

namespace PerfIssue.RC2
{
    class Program
    {
        static void Main(string[] args)
        {
            Console.WriteLine("RC2");
            using (var context = new TestContext())
            {
                Console.WriteLine("Creating database...");
                context.Database.EnsureDeleted();
                context.Database.EnsureCreated();

                Console.WriteLine("Inserting data...");
                
                context.Products.Add(new Products
                {
                    SiteId = 1,
                    Doc = new Documents
                    {
                    }
                });
                context.SaveChanges();
            }

            using (var context = new TestContext())
            {
                var productId = 1;
                var siteId = 1;
                var count = 0;

                Console.WriteLine("Starting test...");
                var end = DateTime.Now.AddSeconds(30);

                while (DateTime.Now < end)
                {
                    var query = from p in context.Products
                                where p.Id == productId && p.SiteId == siteId
                                select p;

                    query = query.Include(b => b.Doc);

                    var x = query.ToList();

                    count++;
                }

                Console.WriteLine($"Result: {count} queries");
                Console.ReadLine();
            }
        }
    }

    public class TestContext : DbContext
    {
        public DbSet<Products> Products { get; set; }
        public DbSet<Documents> Documents { get; set; }

        protected override void OnConfiguring(DbContextOptionsBuilder optionsBuilder)
        {
            optionsBuilder.UseSqlServer(@"Server=(localdb)\mssqllocaldb;Database=TestDb;Trusted_Connection=True;");
        }
    }


    public partial class Products
    {
        public int Id { get; set; }
        public int SiteId { get; set; }

        public int? DocId { get; set; }
        public virtual Documents Doc { get; set; }
    }

    public partial class Documents
    {
        public int Id { get; set; }

        public virtual ICollection<Products> Products { get; set; } = new HashSet<Products>();
    }
}
```

_project.json_
```
{
  "version": "1.0.0-*",
  "buildOptions": {
    "emitEntryPoint": true
  },

  "dependencies": {
    "Microsoft.EntityFrameworkCore": "1.0.0-rc2-final",
    "Microsoft.EntityFrameworkCore.SqlServer": "1.0.0-rc2-final"
  },

  "frameworks": {
    "dnx451": {
      "dependencies": {
        "Microsoft.CSharp": "4.0.1-beta-23516",
        "System.Collections": "4.0.11-beta-23516",
        "System.Linq": "4.0.1-beta-23516",
        "System.Runtime": "4.0.21-beta-23516",
        "System.Threading": "4.0.11-beta-23516"
      }
    }
  }
}
```

_Results_

38306 queries

![2016-07-11 12_06_20-lrpublisherweb - remote desktop connection](https://cloud.githubusercontent.com/assets/17545705/16727464/7908fd72-4761-11e6-8a34-b895e0dde2f6.png)


Ran the tests multiple times and the results were always consistent.

**Comment by rowanmiller** [comment]
Confirmed that I see the same results running between RC1 and RTM. Looks like a bug in RC1 that was causing invalid data to be saved to the database in my repro (hence no data to return in the query). The repro supplied by @janiceattard has a workaround for this.

Some data points to consider:
* What's interesting about this test is that it runs the same tracking query a lot of times on the same context instance... so identity resolution is probably one of the main code paths being exercised here.
* The RC1 numbers are similar to what you get with EF6.1.3, so the RTM numbers are slower. 
* Our Include benchmarks, which use a fresh context instance for each query, show EF Core as ~50% faster than EF6.x. 
* Removing the Include results in ~35% more throughput on RTM (compared to RC1)... so the slowdown is definitely related to the Include.

This may be something we want to profile, as the hot path could be a quick win for a perf improvement.

**Comment by rowanmiller** [comment]
@janiceattard do you have a real application scenario that this is impacting? In general, Include is faster than RC1 and EF6.x, and it seems like this "repeatedly querying the same data that is already tracked by the context (therefore querying the database, but then just returning the entity instances already in-memory)" isn't a scenario that a real app would typically do.

**Comment by janiceattard** [comment]
@rowanmiller in fact, in our application we are not continually querying the same data - in most of the cases the productids are different. So it's not just the case when querying the same data repeatedly.

We tested with a fresh context for each query using the below code:

```
var end = DateTime.Now.AddSeconds(30);
var count = 0;

while (DateTime.Now < end)
{

                using (var context = new TestContext())
                {
                    var productId = 1;
                    var siteId = 1;

                    var query = from p in context.Products
                                where p.Id == productId && p.SiteId == siteId
                                select p;

                    query = query.Include(b => b.Doc);

                    var x = query.ToList();

                    count++;
                }
}

Console.WriteLine($"Result: {count} queries");
Console.ReadLine();
```

But both RC2 and RTM are still slower than RC1:

![2016-07-25 17_19_15-c__windows_system32_cmd exe](https://cloud.githubusercontent.com/assets/17545705/17107789/05a64444-5291-11e6-9ac3-c242a6325664.png)

We also tried using `var x = query.AsNoTracking().ToList();` as the issue seems to be with data already being tracked - there was some improvement but not much.

Would you be able to provide a code listing where RTM is faster than RC1 to check if we replicate?

**Comment by rowanmiller** [comment]
Ok, finally got to the bottom of this one... it's when the result count is small. Comparing to RC1 isn't really something we are concerned with (since a lot of things were faster because the product was only partially implemented) but I used EF6.x as a baseline (which we do care about) and saw the same relative results. I've opened https://github.com/aspnet/EntityFramework/issues/6181 to track this specific issue now that it is isolated down.
